### ISSUE TITLE
Logistic Regression Defunct L1 & L2 penalty. Bug & sugguested fix

### ISSUE THREAD

**Issue by cqgz21** [HazyResearch/snorkel#872]
L1 & L2 penalty are not actually taken into consideration by the tensorflow optimizer. Regardless of what value the l1_penalty takes, the optimized weights are not affected at all. Try run the following:

import numpy as np
from snorkel.learning import LogisticRegression
train_marginals = np.random.uniform(0,0.3,size = 100)
X = np.random.normal(size = (100,1000))

model = LogisticRegression()
model.train(X,train_marginals,n_epochs=1000, print_freq=100, l1_penalty=0,lr = 0.01)
wts = model.get_weights()

model2 = LogisticRegression()
model2.train(X,train_marginals,n_epochs=1000, print_freq=100, l1_penalty=10,lr = 0.01)
wts2 = model2.get_weights()

sum(wts[0] - wts2[0])

I believe doing the following would be a makeshift fix:
Go to /snorkel/learning/disc_learning.py, copy lines 83-84 to clipboard and comment them out. Then go to /snorkel/learning/disc_models/logistic_regression.py ; paste the 2 lines at line at line 76 (inside the _build_training_ops after the "l1" "l2" if blocks.

I believe the issue is with the class inheritance. In the old way, MyObject.optimizer is defined once using MyObject.loss. Then, MyObject.loss is set to add the penalty. Aparently this order does not work. We should first define MyObject.loss AND add the penalty, then define MyObject.optimizer using MyObject.loss.

**Comment by stephenbach** [comment]
Hi! Thanks for the report!

Executing this code on 0.7.0-alpha, I get
```
[LogisticRegression] Training model
[LogisticRegression] n_train=100  #epochs=1000  batch size=100
[LogisticRegression] Epoch 0 (0.02s)	Average loss=1.405735
[LogisticRegression] Epoch 100 (0.11s)	Average loss=0.388082
[LogisticRegression] Epoch 200 (0.19s)	Average loss=0.387633
[LogisticRegression] Epoch 300 (0.29s)	Average loss=0.387630
[LogisticRegression] Epoch 400 (0.38s)	Average loss=0.387630
[LogisticRegression] Epoch 500 (0.46s)	Average loss=0.387630
[LogisticRegression] Epoch 600 (0.55s)	Average loss=0.387630
[LogisticRegression] Epoch 700 (0.64s)	Average loss=0.387630
[LogisticRegression] Epoch 800 (0.73s)	Average loss=0.387630
[LogisticRegression] Epoch 900 (0.82s)	Average loss=0.387630
[LogisticRegression] Epoch 999 (0.91s)	Average loss=0.387630
[LogisticRegression] Training done (0.91s)
[LogisticRegression] Training model
[LogisticRegression] n_train=100  #epochs=1000  batch size=100
[LogisticRegression] Epoch 0 (0.01s)	Average loss=777.950623
[LogisticRegression] Epoch 100 (0.12s)	Average loss=12.006347
[LogisticRegression] Epoch 200 (0.22s)	Average loss=11.425220
[LogisticRegression] Epoch 300 (0.32s)	Average loss=11.224578
[LogisticRegression] Epoch 400 (0.43s)	Average loss=10.959377
[LogisticRegression] Epoch 500 (0.53s)	Average loss=11.042912
[LogisticRegression] Epoch 600 (0.64s)	Average loss=11.138163
[LogisticRegression] Epoch 700 (0.74s)	Average loss=10.659776
[LogisticRegression] Epoch 800 (0.86s)	Average loss=11.119858
[LogisticRegression] Epoch 900 (0.96s)	Average loss=10.854328
[LogisticRegression] Epoch 999 (1.07s)	Average loss=10.700669
[LogisticRegression] Training done (1.07s)
-0.26523290685145184
```
Looking at the weights themselves, wts2 is definitely much smaller in magnitude. Similar results for l2.

Is it possible that this is tensorflow version specific? I'm using 1.1.0.

**Comment by ajratner** [comment]
Closing this for now, feel free to re-open!
